### PR TITLE
Mitigate slow getnameinfo(3) problem on macOS

### PR DESCRIPTION
**Pull request by masaori335**
#2565

### CODE REVIEW

File: iocore/utils/Machine.cc [hunk 1]
```diff
@@ -175,8 +176,14 @@ Machine::Machine(char const *the_hostname, sockaddr const *addr)
...
           insert_id(ip_strbuf);
-          if (getnameinfo(ifip, ats_ip_size(ifip), localhost, sizeof(localhost) - 1, nullptr, 0, 0) == 0) {
-            insert_id(localhost);
+
+          // XXX: #2565 mitigate slow getnameinfo(3) problem on macOS
+          bool parent_proxy_routing_enable = false;
+          REC_ReadConfigInteger(parent_proxy_routing_enable, "proxy.config.http.parent_proxy_routing_enable");
+          if (parent_proxy_routing_enable) {
+            if (getnameinfo(ifip, ats_ip_size(ifip), localhost, sizeof(localhost) - 1, nullptr, 0, 0) == 0) {
+              insert_id(localhost);
```
Comment (masaori335) [inline iocore/utils/Machine.cc]: @jrushford Is `machine_id_strings` table used for anywhere else ParentSelection?

**Comment by jrushford** [comment]
@masaori335<https://github.com/masaori335>? currently only parent selection uses this but, it's designed to provide some self detection that will probably be used by next hop.


If you run a google search, Mac OS seems to have several issues with dns lookups, from a google search, particularly if it's a reverse lookup and there is no reverse definition in DNS.  Have you tried adding PTR records for each IP address on your nic(s) or adding an entry in /etc/hosts for each ip?


I think later that this PR would cause issue when using the self detection in next hop.


________________________________

**Comment by masaori335** [comment]
Thank you for advice, I'll try them and dig more. And if I couldn't find anything, I'll add new config variable for this.

**Comment by zwoop** [comment]
[approve ci fedora]

**Comment by jrushford** [comment]
@zwoop, I see your point but, this self detection api will probably be used very soon with next hop and i was hoping we could get to the bottom of the issue @masaori335 is having.  if it's a local configuration problem then perhaps nothing needs to be done.

**Comment by zwoop** [comment]
[approve ci fedora]

**Comment by masaori335** [comment]
I tried add entries in /etc/hosts, change hostname ... but I couldn't find fix or root cause :(

I have 2 slow IPv6 link-local address. And it looks like `getnameinfo` waits until timeout (5 seconds) for each.
And I marked an interface of them down. Then I noticed TouchID is using that interface, so I don't want to do that.

Our consensus is merge this for a workaround, right? I’ll leave the issue open.

**Comment by jrushford** [comment]
Ah ha.  Instead of merging this what do you think of modifying the code to skip link-local addresses, ipv4 or ipv6?   Shouldn’t be trying to resolve those anyway.

**Comment by jrushford** [comment]
Masaori,

Do you want to try this change?

178           if (! ats_is_ip_linklocal(ifip) && getnameinfo(ifip, ats_ip_size(ifip), localhost, sizeof(localhost) - 1, nullptr, 0, 0) == 0) {
179             insert_id(localhost);
180           }

From: John Rushford <John_Rushford@cable.comcast.com>
Date: Thursday, October 19, 2017 at 7:06 PM
To: apache/trafficserver <reply@reply.github.com>, apache/trafficserver <trafficserver@noreply.github.com>
Cc: Mention <mention@noreply.github.com>
Subject: Re: [apache/trafficserver] Mitigate slow getnameinfo(3) problem on macOS (#2664)

Ah ha.  Instead of merging this what do you think of modifying the code to skip link-local addresses, ipv4 or ipv6?   Shouldn’t be trying to resolve those anyway.

**Comment by masaori335** [comment]
Oh, can we skip all link local address? I thought it makes problems in some cases.

**Pull request closed without merging**